### ISSUE TITLE
Clarify how failures abort affect subsequent commands in a stage

### ISSUE THREAD

**Issue by Lekensteyn** [travis-ci/docs-travis-ci-com#2767]
For readability and maintaining timing info for a command, it is desirable to specify commands in a pipeline on separate lines. For example:

```yaml
before_script:
- sudo apt-get update
- sudo apt-get install foo
script:
- make
- make test
```

In this example, when the first or second command fails, it will avoid executing all others.

However when "make" fails, it will still cause "make test" to be executed. This was very surprising, but it is apparently how [build stages](https://docs.travis-ci.com/user/build-stages) work. I don't know if this behavior is documented anywhere, but it appears that some stages (such as before_install and before_script) always abort on the first failure, but others (such as script and after_script) do not.

Internally, it appears that stages with `assert: true` achieve the desired behavior, and that `script` has `assert: false`: https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/stages.rb

As workaround I now use:

```yaml
script:
- make || travis_terminate
- make test
```

but this function is not documented although it has been there since 2012 via travis-ci/travis-build@b472ee51. The documentation does mention `travis_terminate`, but that seems to avoid some logging done by `travis_assert`. Additionally, the current implementation of `travis_terminate` fails since it ends up calling `exit ""` when the first parameter is missing, so right now you have to use `travis_terminate $?` to avoid that.

Another data point: https://github.com/travis-ci/docs-travis-ci-com/tree/master/user/build-stages/share-docker-image.md also has a list of commands, but it looks like it expects the same "exit on the first failing command", otherwise subsequent `docker` commands will fail anyway.

If the current "script" command cannot be made more obvious, then the documentation should somehow be clarified with some recommendations.

**Comment by BanzaiMan** [comment]
This is documented in https://docs.travis-ci.com/user/job-lifecycle#customizing-the-build-phase; we welcome your suggestions on how to improve the treatment of the subject and make it easier to find.

---

> However when "make" fails, it will still cause "make test" to be executed. This was very surprising, but it is apparently how build stages work. 

This is simply untrue. Here, `make` and `make test` are run within a single job, over which Build Stage has no control.

The implementation details of how these commands in `travis-build` is executed predates the more recent feature we call "Build Stages"; they are similarly named, but what's in `travis-build` is not referred to as "stages" in the documentation.

The "surprise" has been discussed before; https://github.com/travis-ci/travis-ci/issues/1066

**Comment by Lekensteyn** [comment]
> This is documented [..]

Oh wow, I missed that note. I think I was looking at https://docs.travis-ci.com/user/customizing-the-build/ and https://docs.travis-ci.com/user/common-build-problems/ for "fail" and "exit". The suggested behavior is to combine everything in a single pipeline or script file, but that may be undesirable for several reasons. For me, having separate items enables Travis to show running times for each individual step. What would help is additional documentation to ensure the desired behavior is achieved (namely, fail early).

Regarding the "build stages", I might have linked the wrong details with the observed behavior for the "script" block, I am not intimately familiar with the internal details. But it matches what I observed.

Some concrete suggestions:

 - Introduce a command/option to ensure that commands fail ASAP. E.g.:
```
script:
 - travis_ensure_fail_early
 - make
 - make test
```
- Document how to properly abort the script block, e.g. using `- make || travis_assert` or similar
- Change the default behavior to make it work as expected (i.e. fail early). This is consistent with how other steps work.
- If subsequent commands are ran due to an earlier failure, consider outputting a reference to the documentation that explains this "surprise". This would have saved me some time and allowed me to address the problems months ago instead of recently when I finally got around investigating the weird behavior.
